### PR TITLE
Fixes #7778: Add `import.meta.dirname`, `import.meta.filename` for Node compatibility

### LINKED ISSUE #7778: Node compat: Add `import.meta.dirname` and `import.meta.filename`
### What version of Bun is running?

1.0.18+36c316a24

### What platform is your computer?

Linux 5.15.0-1048-oracle aarch64 aarch64

### What steps can reproduce the bug?

```js                                                                                    
// test.mjs
console.log("dir:", import.meta.dir);
console.log("dirname:", import.meta.dirname);
console.log("file:", import.meta.file);
console.log("filename:", import.meta.filename);
```

### What is the expected behavior?

```
$ node test.mjs
dir: undefined
dirname: /home/ubuntu
file: undefined
filename: /home/ubuntu/test.mjs
```

### What do you see instead?

```
$ bun run test.mjs
dir: /home/ubuntu
dirname: undefined
file: test.mjs
filename: undefined
```

### Additional information

Since Node.js v21.2.0, `import.meta.dirname` and `import.meta.filename` are added. https://github.com/nodejs/node/pull/48740

`import.meta.dirname` can be added to Bun as an alias for `import.meta.dir`.

However, while `import.meta.filename` in Node.js includes the directory but `import.meta.file` in Bun does not, `import.meta.filename` would not be just an alias.
`import.meta.filename` equals `${import.meta.dir}/${import.meta.file}`.

I think the behavior might cause some confusion, but I believe that adding these properties would improve compatibility with Node.js.

Node.js docs: https://nodejs.org/api/esm.html#importmeta
Bun docs: https://bun.sh/docs/api/import-meta

### PR DESCRIPTION
**Pull request by knightspore**
### What does this PR do?

This PR fixes #7778 by adding `import.meta.dirname` and `import.meta.filename`, replicating [new additions to Node v.21.2.0](https://github.com/nodejs/node/pull/48740)

As suggested in the issue above, `import.meta.dirname` has the same behaviour as Bun's existing `import.meta.dir`.

Bun's `import.meta.file` does not prepend the path to the filename - so for `import.meta.filename` we return a value equivalent to `${import.meta.dir}/${import.meta.file}` as suggested in the original issue. 

There are some slight changes on the desired output listed in #7778 since none of these `import.meta` properties should currently return `undefined`:

|Meta Property|Expected (Node)| Seen Instead (Bun)|New Behaviour (Bun)|
|--|--|--|--|
|`dir`|`undefined`|`/home/ubuntu`|`/home/ubuntu`|
|`dirname`|`/home/ubuntu`|`undefined`|`/home/ubuntu`|
|`file`|`undefined`|`test.mjs`|`test.mjs`|
|`filename`|`/home/ubuntu/test.mjs`|`undefined`|`/home/ubuntu/test.mjs`|
- [x] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [X] Code changes

### How did you verify your code works?

- Regression tests
- Updating tests in `test/js/bun/resolve/import-meta.test.js` to cover new cases.

- [X] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [X] I included a test for the new code, or an existing test covers it
- [X] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [X] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test/js/bun/resolve/import-meta.test.js test/regression/issue/07778.test.ts`)

### CODE REVIEW

File: packages/bun-types/globals.d.ts [hunk 1]
```diff
@@ -302,10 +302,20 @@ interface ImportMeta {
    * Does not have a trailing slash
    */
   readonly dir: string;
+  /**
+   * Absolute path to the directory containing the source file. (alias for `dir` to match Node.js)
+   *
+   * Does not have trailing slash
+   */
+  readonly dirname: string;
```
Comment (paperclover) [inline packages/bun-types/globals.d.ts:310]: i think you opened this before our big `@types/bun` PR. but can you remove these type declarations. they should be provided by `@types/node`

File: src/bun.js/bindings/ImportMetaObject.cpp [hunk 2]
```diff
@@ -641,6 +683,24 @@ void ImportMetaObject::finishCreation(VM& vm)
 
         init.set(jsString(init.vm, filename));
     });
+    this->filenameProperty.initLater([](const JSC::LazyProperty<JSC::JSObject, JSC::JSString>::Initializer& init) {
```
Comment (paperclover) [inline src/bun.js/bindings/ImportMetaObject.cpp:686]: this lazy property is identical to `this->pathProperty` please remove it.

File: src/bun.js/bindings/ImportMetaObject.cpp [hunk 3]
```diff
@@ -615,6 +635,28 @@ void ImportMetaObject::finishCreation(VM& vm)
 
         init.set(jsString(init.vm, dirname));
     });
+    this->dirnameProperty.initLater([](const JSC::LazyProperty<JSC::JSObject, JSC::JSString>::Initializer& init) {
```
Comment (paperclover) [inline src/bun.js/bindings/ImportMetaObject.cpp:638]: this lazy property is identical to `this->dirProperty` please remove it.

File: src/bun.js/bindings/ImportMetaObject.cpp [hunk 4]
```diff
@@ -489,7 +507,9 @@ static const HashTableValue ImportMetaObjectPrototypeValues[] = {
...
     { "url"_s, static_cast<unsigned>(JSC::PropertyAttribute::ReadOnly | JSC::PropertyAttribute::CustomAccessor | PropertyAttribute::DontDelete), NoIntrinsic, { HashTableValue::GetterSetterType, jsImportMetaObjectGetter_url, 0 } },
     { "dir"_s, static_cast<unsigned>(JSC::PropertyAttribute::ReadOnly | JSC::PropertyAttribute::CustomAccessor | PropertyAttribute::DontDelete), NoIntrinsic, { HashTableValue::GetterSetterType, jsImportMetaObjectGetter_dir, 0 } },
+    { "dirname"_s, static_cast<unsigned>(JSC::PropertyAttribute::ReadOnly | JSC::PropertyAttribute::CustomAccessor | PropertyAttribute::DontDelete), NoIntrinsic, { HashTableValue::GetterSetterType, jsImportMetaObjectGetter_dir, 0 } },
     { "file"_s, static_cast<unsigned>(JSC::PropertyAttribute::ReadOnly | JSC::PropertyAttribute::CustomAccessor | PropertyAttribute::DontDelete), NoIntrinsic, { HashTableValue::GetterSetterType, jsImportMetaObjectGetter_file, 0 } },
+    { "filename"_s, static_cast<unsigned>(JSC::PropertyAttribute::ReadOnly | JSC::PropertyAttribute::CustomAccessor | PropertyAttribute::DontDelete), NoIntrinsic, { HashTableValue::GetterSetterType, jsImportMetaObjectGetter_filename, 0 } },
```
Comment (paperclover) [inline src/bun.js/bindings/ImportMetaObject.cpp:512]:
```suggestion
    { "filename"_s, static_cast<unsigned>(JSC::PropertyAttribute::ReadOnly | JSC::PropertyAttribute::CustomAccessor | PropertyAttribute::DontDelete), NoIntrinsic, { HashTableValue::GetterSetterType, jsImportMetaObjectGetter_path, 0 } },
```

File: src/bun.js/bindings/ImportMetaObject.cpp [hunk 5]
```diff
@@ -454,6 +456,14 @@ JSC_DEFINE_CUSTOM_GETTER(jsImportMetaObjectGetter_dir, (JSGlobalObject * globalO
 
     return JSValue::encode(thisObject->dirProperty.getInitializedOnMainThread(thisObject));
 }
+JSC_DEFINE_CUSTOM_GETTER(jsImportMetaObjectGetter_dirname, (JSGlobalObject * globalObject, JSC::EncodedJSValue thisValue, PropertyName propertyName))
```
Comment (paperclover) [inline src/bun.js/bindings/ImportMetaObject.cpp:459]: remove this function

File: src/bun.js/bindings/ImportMetaObject.cpp [hunk 6]
```diff
@@ -462,6 +472,14 @@ JSC_DEFINE_CUSTOM_GETTER(jsImportMetaObjectGetter_file, (JSGlobalObject * global
 
     return JSValue::encode(thisObject->fileProperty.getInitializedOnMainThread(thisObject));
 }
+JSC_DEFINE_CUSTOM_GETTER(jsImportMetaObjectGetter_filename, (JSGlobalObject * globalObject, JSC::EncodedJSValue thisValue, PropertyName propertyName))
```
Comment (paperclover) [inline src/bun.js/bindings/ImportMetaObject.cpp:475]: remove this function

**Review by paperclover — CHANGES_REQUESTED** [review]
great to achieve node compatibility, but ouch now we have duplicate properties.

instead of using new LazyProperty for these properties, we can simply reuse the existing ones, even pointing to the same getter implementations.

File: src/bun.js/bindings/ImportMetaObject.cpp (hunk 4 above)
Comment (Jarred-Sumner) [reply src/bun.js/bindings/ImportMetaObject.cpp:512]: we probably need to check if it's readonly in node

**Comment by Jarred-Sumner** [comment]
Superseded by #8127

**Pull request closed without merging**